lib/os/cbprintf_packaged.c

```diff
@@ -0,0 +1,506 @@
+	u.__ap.fp_offset = (6 * 8 + 16 * 16);
+
+	return cbvprintf(out, ctx, fmt, u.ap);
+}
+
+#define VA_STACK_MIN_ALIGN	8
+
+#else
+/*
+ * Default implementation shared by many architectures like
+ * 32-bit ARM and Intel.
+ *
+ * We assume va_list is a simple pointer.
+ */
+
+BUILD_ASSERT(sizeof(va_list) == sizeof(void *),
+	     "architecture specific support is needed");
+
+static int cbprintf_via_va_list(cbprintf_cb out, void *ctx,
+				const char *fmt, void *buf)
+{
+	union {
+		va_list ap;
+		void *ptr;
+	} u;
+
+	u.ptr = buf;
+
+	return cbvprintf(out, ctx, fmt, u.ap);
+}
+
+#endif
+
+/*
+ * Special alignment cases
+ */
+
+#if defined(__i386__)
+/* those are not natually aligned on the stack */
+#define VA_STACK_ALIGN_LONGLONG	4
+#define VA_STACK_ALIGN_DOUBLE	4
+#define VA_STACK_ALIGN_INTMAX	4
+#endif
+
+#if defined(__riscv)
+#define VA_STACK_MIN_ALIGN	(__riscv_xlen / 8)
+#endif
+
+/*
+ * Default alignment values if not specified by architecture config
+ */
+
+#ifndef VA_STACK_MIN_ALIGN
+#define VA_STACK_MIN_ALIGN	1
+#endif
+
+#ifndef VA_STACK_ALIGN_LONGLONG
+#define VA_STACK_ALIGN_LONGLONG	__alignof__(long long)
+#endif
+
+#ifndef VA_STACK_ALIGN_DOUBLE
+#define VA_STACK_ALIGN_DOUBLE	__alignof__(double)
+#endif
+
+#ifndef VA_STACK_ALIGN_INTMAX
+#define VA_STACK_ALIGN_INTMAX	__alignof__(intmax_t)
+#endif
+
+
+int cbvprintf_package(void *packaged, size_t *len,
+		      const char *fmt, va_list ap)
+{
+	unsigned int align, size, i, s_idx = 0;
+	uint8_t str_ptr_pos[16];
+	char *s, *buf = packaged, *buf0 = buf;
+	bool parsing = false;
+
+	/*
+	 * Let's make sure we may store at least the fmt pointer,
+	 * the arg list size and the number of appended strings.
+	 * When buf0 is NULL then we compute the needed length instead.
+	 */
+	if (buf0 && sizeof(char *) + sizeof(int) + 1 > *len) {
+		return -ENOSPC;
+	}
+
+	/* Let's start with the format string itself */
+	str_ptr_pos[s_idx++] = 0;
+	if (buf0) {
+		*(const char **)buf = fmt;
+	} else {
+		/*
+		 * Add the size of the string, the '\0'
+		 * and the pointer position prefix.
+		 */
+		*len = strlen(fmt) + 1 + 1;
+		/*
+		 * And let's also account the string count location here too
+		 * since we'll quit early.
+		 */
+		*len += 1;
+	}
+	buf += sizeof(char *);
+
+	/*
+	 * Then make room to store the arg list size. We store it here
+	 * as it might possibly have a smaller alignment than the fmt pointer.
+	 */
+	buf += sizeof(int);
+
+	/* Scan the format string */
+	for (; *fmt; fmt++) {
+		if (!parsing) {
+			if (*fmt == '%') {
+				parsing = true;
+				align = __alignof__(int);
+				size = sizeof(int);
+			}
+			continue;
+		}
+		switch (*fmt) {
+		case '%':
+			parsing = false;
+			continue;
+
+		case '#':
+		case '-':
+		case '+':
+		case ' ':
+		case '0':
+		case '1':
+		case '2':
+		case '3':
+		case '4':
+		case '5':
+		case '6':
+		case '7':
+		case '8':
+		case '9':
+		case '.':
+		case 'h':
+			continue;
+
+		case '*':
+			break;
+
+		case 'l':
+			if (fmt[1] == 'l') {
```

Comment:
The `l` and `ll` length specifiers mean something different for integral, character, and floating point types.